File: appcore/src/main/java/mobi/cangol/mobile/http/route/RouteHttpClient.java
/**
 * Copyright (c) 2013 Cangol
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.http.route;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import mobi.cangol.mobile.service.PoolManager;
import okhttp3.Call;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class RouteHttpClient {
    public static final String TAG = "RouteHttpClient";
    private static final int DEFAULT_CONNECT_TIMEOUT = 30 * 1000;
    private static final int DEFAULT_READ_TIMEOUT = 30 * 1000;
    private static final int DEFAULT_WRITE_TIMEOUT = 30 * 1000;
    private static final int DEFAULT_MAX = 3;
    private static Map<Object, List<WeakReference<Future<?>>>> requestMap = new WeakHashMap<>();
    private OkHttpClient httpClient;
    private PoolManager.Pool threadPool;

    /**
     * 构造实例
     */
    public RouteHttpClient() {

        httpClient = new OkHttpClient.Builder()
                .retryOnConnectionFailure(true)
                .followRedirects(true)
                .followSslRedirects(true)
                .readTimeout(DEFAULT_READ_TIMEOUT, TimeUnit.MILLISECONDS)
                .connectTimeout(DEFAULT_CONNECT_TIMEOUT, TimeUnit.MILLISECONDS)
                .writeTimeout(DEFAULT_WRITE_TIMEOUT, TimeUnit.MILLISECONDS)
                .build();
        threadPool = PoolManager.buildPool(TAG, DEFAULT_MAX);

    }

    /**
     * 发起请求
     *
     * @param tag
     * @param url
     * @param params
     * @param responseHandler
     * @param host
     */
    public void send(Object tag, String url, Map<String, String> params, RouteResponseHandler responseHandler, String... host) {
        Request request = null;
        if (params != null) {
            final FormBody.Builder requestBodyBuilder = new FormBody.Builder();
            for (final ConcurrentHashMap.Entry<String, String> entry : params.entrySet()) {
                requestBodyBuilder.add(entry.getKey(), entry.getValue());
            }
            request = new Request.Builder()
                    .tag(tag)
                    .url(url)
                    .post(requestBodyBuilder.build())
                    .build();
        } else {
            request = new Request.Builder()
                    .tag(tag)
                    .url(url)
                    .build();
        }
        sendRequest(httpClient, request, responseHandler, tag, host);
    }

    private Request getNewRequest(Request request, String host) {
        final String hostStr = request.url().url().getHost();
        final String urlStr = request.url().url().toString().replace(hostStr, host);

        return new Request.Builder()
                .tag(request.tag())
                .url(urlStr)
                .build();
    }

    protected void sendRequest(OkHttpClient client, Request uriRequest, RouteResponseHandler responseHandler, Object context, String... host) {

        final Future<?> request = threadPool.submit(new HttpRequestTask(client, uriRequest, responseHandler, host));
        if (context != null) {
            // Add request to request map
            List<WeakReference<Future<?>>> requestList = requestMap.get(context);
            if (requestList == null) {
                requestList = new LinkedList<>();
                requestMap.put(context, requestList);
            }
            requestList.add(new WeakReference<Future<?>>(request));
        }
    }

    /**
     * 取消请求
     *
     * @param tag
     * @param mayInterruptIfRunning
     */
    public void cancelRequests(Object tag, boolean mayInterruptIfRunning) {
       final List<WeakReference<Future<?>>> requestList = requestMap.get(tag);
        if (requestList != null) {
            for (final WeakReference<Future<?>> requestRef : requestList) {
                final Future<?> request = requestRef.get();
                if (request != null) {
                    request.cancel(mayInterruptIfRunning);
                }
            }
        }
        requestMap.remove(tag);
        for (final Call call : httpClient.dispatcher().queuedCalls()) {
            if (call.request().tag().equals(tag)) {
                call.cancel();
            }
        }
        for (final Call call : httpClient.dispatcher().runningCalls()) {
            if (call.request().tag().equals(tag)) {
                call.cancel();
            }
        }
    }

    class HttpRequestTask implements Runnable {
        private final RouteResponseHandler responseHandler;
        private OkHttpClient client;
        private Request request;
        private String[] host;

        public HttpRequestTask(OkHttpClient client, Request request, RouteResponseHandler responseHandler, String... host) {
            this.client = client;
            this.request = request;
            this.responseHandler = responseHandler;
            this.host = host;
        }

        @Override
        public void run() {
            if (!Thread.currentThread().isInterrupted()) {
                responseHandler.sendStartMessage();
                int exec = 0;
                while (exec < host.length) {
                    try {
                        request = getNewRequest(request, host[exec]);
                        exec++;
                        final Response response = client.newCall(request).execute();
                        if (!Thread.currentThread().isInterrupted()) {
                            if (responseHandler != null) {
                                if (responseHandler.sendResponseMessage(response)) {
                                    break;
                                } else {
                                    //
                                }
                            }
                        } else {
                            break;
                        }
                    } catch (IOException e) {
                        if (exec >= host.length) {
                            responseHandler.sendFailureMessage(e, "IOException");
                            break;
                        }
                    }
                }
            }
        }
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/http/route/RouteHttpClient.kt
/**
 * Copyright (c) 2013 Cangol
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.http.route

import mobi.cangol.mobile.service.PoolManager
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

class RouteHttpClient {
    private val httpClient: OkHttpClient
    private val threadPool: PoolManager.Pool

    /**
     * 构造实例
     */
    init {

        httpClient = OkHttpClient.Builder()
                .retryOnConnectionFailure(true)
                .followRedirects(true)
                .followSslRedirects(true)
                .readTimeout(DEFAULT_READ_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                .connectTimeout(DEFAULT_CONNECT_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                .writeTimeout(DEFAULT_WRITE_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                .build()
        threadPool = PoolManager.buildPool(TAG, DEFAULT_MAX)

    }

    /**
     * 发起请求
     *
     * @param tag
     * @param url
     * @param params
     * @param responseHandler
     * @param host
     */
    fun send(tag: Any, url: String, params: Map<String, String>?, responseHandler: RouteResponseHandler, vararg host: String) {
        var request: Request?
        if (params != null) {
            val requestBodyBuilder = FormBody.Builder()
            for ((key, value) in params) {
                requestBodyBuilder.add(key, value)
            }
            request = Request.Builder()
                    .tag(tag)
                    .url(url)
                    .post(requestBodyBuilder.build())
                    .build()
        } else {
            request = Request.Builder()
                    .tag(tag)
                    .url(url)
                    .build()
        }
        sendRequest(httpClient, request, responseHandler, tag, *host)
    }

    private fun getNewRequest(request: Request, host: String): Request {
        val hostStr = request.url().url().host
        val urlStr = request.url().url().toString().replace(hostStr, host)

        return Request.Builder()
                .tag(request.tag())
                .url(urlStr)
                .build()
    }

    protected fun sendRequest(client: OkHttpClient, uriRequest: Request?, responseHandler: RouteResponseHandler, context: Any?, vararg host: String) {

        val request = threadPool.submit(HttpRequestTask(client, uriRequest, responseHandler, *host))
        if (context != null) {
            // Add request to request map
            var requestList: MutableList<WeakReference<Future<*>>>? = requestMap[context]
            if (requestList == null) {
                requestList = LinkedList()
                requestMap[context] = requestList
            }
            requestList.add(WeakReference(request))
        }
    }

    /**
     * 取消请求
     *
     * @param tag
     * @param mayInterruptIfRunning
     */
    fun cancelRequests(tag: Any, mayInterruptIfRunning: Boolean) {
        val requestList = requestMap[tag]
        if (requestList != null) {
            for (requestRef in requestList) {
                val request = requestRef.get()
                request?.cancel(mayInterruptIfRunning)
            }
        }
        requestMap.remove(tag)
        for (call in httpClient.dispatcher().queuedCalls()) {
            if (call.request().tag() == tag) {
                call.cancel()
            }
        }
        for (call in httpClient.dispatcher().runningCalls()) {
            if (call.request().tag() == tag) {
                call.cancel()
            }
        }
    }

    internal inner class HttpRequestTask(private val client: OkHttpClient, private var request: Request?, private val responseHandler: RouteResponseHandler?, vararg host: String) : Runnable {
        private val host: Array<out String>?

        init {
            this.host = host
        }

        override fun run() {
            if (!Thread.currentThread().isInterrupted) {
                responseHandler?.sendStartMessage()
                var exec = 0
                while (exec < host!!.size) {
                    try {
                        request = getNewRequest(request!!, host[exec])
                        exec++
                        val response = client.newCall(request).execute()
                        if (!Thread.currentThread().isInterrupted) {
                                if (responseHandler?.sendResponseMessage(response)!!) {
                                    break
                                } else {
                                    //
                                }
                        } else {
                            break
                        }
                    } catch (e: IOException) {
                        if (exec >= host.size) {
                            responseHandler?.sendFailureMessage(e, "IOException")
                            break
                        }
                    }

                }
            }
        }
    }

    companion object {
        private const val TAG = "RouteHttpClient"
        private const val DEFAULT_CONNECT_TIMEOUT = 30 * 1000
        private const val DEFAULT_READ_TIMEOUT = 30 * 1000
        private const val DEFAULT_WRITE_TIMEOUT = 30 * 1000
        private const val DEFAULT_MAX = 3
        private val requestMap = WeakHashMap<Any, MutableList<WeakReference<Future<*>>>>()
    }
}

